rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+executed end-to-end on an accelerator.
+
+This means that we need a way to lift the preprocessing part of the model in a tf.data pipeline during model training. In `fit`, we can do this automatically.
+In custom training loops, we will expect the user to do it manually (see subsection "Custom training loops").
+
+We propose the addition of two new methods on the `Model` class:
+
+```python
+def get_preprocessing_stage(self):
+    """Retrieves the preprocessing part of the model.
+
+    This is the part of the model that should be executed asynchronously
+    on the device host during training.
+
+    Returns:
+        Instance of `PreprocessingLayer` or `PreprocessingStage`.
+        May be None if the model does not start with preprocessing layers.
+    """
+    pass
+
+def get_main_stage(self):
+    """Retrieves the main processing part of the model.
+
+    This is the part of the model that should be executed
+    on the accelator device.
+
+    Returns:
+        Model instance.
+    """
+```
+
+Thus, for any model that starts with preprocessing layers, the following:
+
+```python
+outputs = model(inputs)
+```
+
+is functionnally equivalent to:
+
+```python
+preprocessed_inputs = model.get_preprocessing_stage()(inputs)
+outputs = model.get_main_stage()(preprocessed_inputs)
+```
+
+
+#### Examples:
+
+Sequential model with a preprocessing layer:
+
+```python
+vectorization = keras.layers.TextVectorization()
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Dense(10, activation='softmax'),
+])
+
+# This is the `vectorization` layer.
+preproc_stage = model.get_preprocessing_stage()
+# model containing the `Dense` layer only.
+main_stage = model.get_main_stage()
+```
+
+Functional model with 2 branches, each with a preprocessing layer:
+
+```python
+normalization_a = layers.Normalization()
+normalization_b = layers.Normalization()
+normalization_a.adapt(data_a)
+normalization_b.adapt(data_b)
+
+input_a = Input(shape_a)
+input_b = Input(shape_b)
+normed_a = normalization_a(input_a)
+normed_b = normalization_b(input_b)
+a = layers.Dense(32)(normed_a)
+b = layers.Dense(32)(normed_b)
+c = layers.concatenate([a, b])
+outputs = layers.Dense(1, activation='sigmoid')(c)
+
+model = Model([input_a, input_b], outputs)
+
+# `PreprocessingStage` instance
+# mapping `[input_a, input_b]` to `[normed_a, normed_b]`
+preproc_stage = model.get_preprocessing_stage()
+
+# Model instance mapping `[normed_a, normed_b]` to `outputs`.
+main_stage = model.get_main_stage()
+```
+
+Subclassed model with a preprocessing layer:
+
+```python
+class MyModel(Model):
+
+    def __init__(self, **kwargs):
+        super(MyModel, self).__init__(**kwargs)
+        self.preproc_layer = layers.Normalization()
+        self.submodel = MySubmodel()
+
+    def call(self, inputs):
+        return self.submodel(self.preproc_layer)
```

Comment:
typo: we forgot inputs should be 
` return self.submodel(self.preproc_layer(inputs))`